### PR TITLE
Support collecting worker metrics in collectMetrics command

### PR DESCRIPTION
**Pull request by ljl1988com**
### What changes are proposed in this pull request?
#14966 : Support collecting worker metrics in collectMetrics command. 

### Why are the changes needed?
Support collecting worker metrics in collectMetrics command

### Does this PR introduce any user facing changes?
None

### CODE REVIEW

File: shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java [hunk 1]
```diff
@@ -63,22 +65,75 @@ public int run(CommandLine cl) throws AlluxioException, IOException {
...
+    outputBuffer.write(metricsResponse);
+    outputBuffer.write("\n");
+
+    // Write to file
+    File outputFile = generateOutputFile(mWorkingDirPath,
+            String.format("%s-master-%s", getCommandName(), i));
+    FileUtils.writeStringToFile(outputFile, metricsResponse);
+  }
+
+  private void workMetrics(StringWriter outputBuffer, int i) throws IOException {
```
Comment (beinan) [inline shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java:123]:
```suggestion
  private void workerMetrics(StringWriter outputBuffer, int i) throws IOException {
```
```suggestion
  private void workMetrics(StringWriter outputBuffer, int i) throws IOException {
```

Comment (beinan) [inline shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java:123]:
```suggestion
  private void workerMetrics(StringWriter outputBuffer, int i) throws IOException {
```

File: shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java [hunk 2]
```diff
@@ -63,22 +65,75 @@ public int run(CommandLine cl) throws AlluxioException, IOException {
...
-              mFsContext.getClusterConf().get(PropertyKey.MASTER_WEB_PORT),
+    // TODO(jiacheng): phase 2 consider outputting partial results in a finally block
+    File outputFile = generateOutputFile(mWorkingDirPath,
+            String.format("%s.txt", getCommandName()));
+    FileUtils.writeStringToFile(outputFile, outputBuffer.toString());
+
+    return 0;
+  }
+
+  private void masterMetrics(StringWriter outputBuffer, int i) throws IOException {
```
Comment (beinan) [inline shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java:88]: the method argument `int i` is unnecessary, am I right?

File: shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java (hunk 1 above)
Comment (beinan) [inline shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java:123]: the method argument `int i` is unnecessary, am I right?

Comment (beinan) [inline shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java:123]: the method argument `int i` is unnecessary, am I right?

**Review by beinan — COMMENTED** [review]
Hello @ljl1988com , thank you for the contribution! It looks good to me except a couple of minor issues.

Another concern is if it would take too much time if there are hundreds of alluxio workers.  Is it possible to add a command argument/flag to enable/disable this feature?

File: shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ljl1988com) [reply shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java:88]: The metrics is invoked 3 times. The interval is 3 seconds. Argument `int i` is used to compose a part of the file name every times.

File: shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java (hunk 1 above)
Comment (ljl1988com) [reply shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java]: Done

File: shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (beinan) [reply shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java:88]: got it, thx!

File: shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java [hunk 3]
```diff
@@ -63,22 +65,75 @@ public int run(CommandLine cl) throws AlluxioException, IOException {
...
+    FileUtils.writeStringToFile(outputFile, metricsResponse);
+  }
+
+  private void workerMetrics(StringWriter outputBuffer, int i) throws IOException {
+    // Generate URL from config properties
+    List<BlockWorkerInfo> workers;
+    try {
+      workers = mFsContext.getCachedWorkers();
+    } catch (UnavailableException e) {
+      String noMasterMsg = "No Alluxio workers available. Skip metrics collection.";
```
Comment (beinan) [inline shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java:129]:
```suggestion
      String noWorkerMsg = "No Alluxio workers available. Skip metrics collection.";
```

**Comment by beinan** [comment]
> @beinan Thanks for your review. PTAL.

will it take too much time if there are hundreds of alluxio workers?  Is it possible to add a command argument/flag to enable/disable the metrics collecting from workers?

File: shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java (hunk 3 above)
Comment (ljl1988com) [reply shell/src/main/java/alluxio/cli/bundler/command/CollectMetricsCommand.java]: done.

**Comment by ljl1988com** [comment]
> 
> will it take too much time if there are hundreds of alluxio workers? Is it possible to add a command argument/flag to enable/disable the metrics collecting from workers?

@beinan Thanks for your review once again. I have added a command flag `only-master` to disable the metrics collecting from workers.

**Comment by ljl1988com** [comment]
I've added command flag. @beinan PTAL. Thanks

File: shell/src/main/java/alluxio/cli/bundler/CollectInfo.java [hunk 4]
```diff
@@ -80,7 +80,7 @@
           + "the cluster.\n"
           + "collectJvmInfo:     collects jstack from the JVMs.\n"
           + "collectLog:         collects the log files under ${ALLUXIO_HOME}/logs/.\n"
-          + "collectMetrics:     collects Alluxio system metrics.\n\n"
+          + "collectMetrics [--only-master]:     collects Alluxio system metrics.\n\n"
```
Comment (beinan) [inline shell/src/main/java/alluxio/cli/bundler/CollectInfo.java:83]: This line is not following the pattern we already have, do you mind add --only-master to line 68~70? looks like we put the flags in the front.  Also, do you think --exclude-worker-metrics or --include-worker-metrics is a better name?

Comment (ljl1988com) [reply shell/src/main/java/alluxio/cli/bundler/CollectInfo.java]: Thanks. I've done. But I think the command argument after command is better.

**Review by beinan — APPROVED** [review]
lgtm, many thanks for the contribution!

**Comment by beinan** [comment]
alluxio-bot, merge this please.

**Pull request merged**